### DISCUSSION TITLE
Content Query where taxonomy is a term OR null/not set? whereTaxonomy combined with whereNull

### DISCUSSION THREAD

**Discussion by sarahessex** [statamic/cms#5466] [Q&A]
Hi, I have two separate content queries which get the results I want:
````
// gets all faqs where segment (a taxonomy) is a certain term
$faqs = Entry::query()
            ->where('collection', 'faqs')
            ->whereTaxonomy("segments::$current_segment")
            ->get();

// gets all faqs where segment taxonomy is null/not set
$faqs = Entry::query()
            ->where('collection', 'faqs')
            ->whereNull("segments")
            ->get();
````

Now I would like to combine these queries into an OR query, so that I get the faqs in the current segment OR that have no segment attached, but I'm having trouble working out the correct parameters.

Some things I've tried, none of which give me the same results merged:

````
// returns an empty set
$faqs = Entry::query()
            ->where('collection', 'faqs')
            ->whereTaxonomy("segments::$current_segment")
            ->orWhereNull("segments")
            ->get();
````

````
// returns only the faqs where segments is null
$faqs = Entry::query()
            ->where('collection', 'faqs')
            ->where(function ($query) {
                // below function does retrieve correct slug, same as above
                $current_segment = $this->getCurrentSegmentSlug();
                $query->whereTaxonomy("segments::$current_segment");
            })
            ->orWhere(function ($query) {
                $query->whereNull("segments");
            })
            ->get();
````

Thanks in advance for any pointer!

**Comment by ryanmitchell** [comment]
```
        $faqs = Entry::query()
            ->where('collection', 'faqs')
            ->where(function ($query) {
                $current_segment = $this->getCurrentSegmentSlug();
                $query->whereTaxonomy("segments::$current_segment")
                ->orWhereNull("segments");
            });
```

should get you there. 

To explain, you are 'grouping' the whereTaxonomy and whereNull, then AND-ing the result with the where collection.

**Reply by sarahessex** [reply]
Thanks for reply, unfortunately this still doesn't give me the entries expected. It only shows me the items where segments is not set/empty.

I've also tried this, but this gives me entries which aren't FAQs

```
$faqs = Entry::query()
            ->where(function ($query) {
                $current_segment = $this->getCurrentSegmentSlug();
                $query->where('collection', 'faqs')
                    ->whereTaxonomy("segments::$current_segment");
            })
            ->orWhere(function ($query) {
                $query->where('collection', 'faqs')
                    ->whereNull("segments");
            });
```

Example of FAQS with segment set...
```
---
id: 3f4f23ad-8bbf-4705-bcc8-0e0f8c0587ff
blueprint: faqs
title: 'An FAQ from Segment One and Two'
segments:
  - segment-one
  - segment-two
```

**Reply by ryanmitchell** [reply]
Try whereTaxonomyIn instead of whereTaxonomy

**Reply by ryanmitchell** [reply]
Oh wait actually ignore that, it won’t work

**Reply by sarahessex** [reply]
Is there a bug here? I would have expected these two queries to be equivalent (even though the bottom one is unnecessarily nested, just as an example)....

```
$faqs = Entry::query()
            ->where('collection', 'faqs')
```
```
$faqs = Entry::query()
            ->where(function ($query) {
                $query->where('collection', 'faqs');
            });
```

But in practice, only the top query gets me only FAQ entries. The bottom query gets me all entries.

It seems like sometimes the nested `where` function gets nothing or has no effect? eg. When doing `$query->where('collection', ...);` or `$query->whereTaxonomy(...)`

This, however, is working as expected for me:

```
$faqs = Entry::query()
            ->where(function ($query) {
                $query->where('title', 'like', '%1%');
            })
            ->orWhere(function ($query) {
                $query->where('title', 'like', '%5%');
            });
```

Am on Statamic 3.2.38.

**Reply by ryanmitchell** [reply]
`whereCollection / where('collection')` and the taxonomy versions aren't the same as querying a column, as they are changing the 'source' of the data. A bit like selecting a table in eloquent. So I'd expect there to be quirks like you are experiencing.

If you want to raise an issue its probably best, and if you had a GitHub repository with examples of code and data so it can be quickly tested that would be helpful.
